components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/internal/IdentityClaimManagementServiceComponent.java

```diff
@@ -208,5 +214,37 @@ protected void unsetIdentityEventService(IdentityEventService identityEventServi
             log.debug("IdentityEventService set in Identity Claim Management bundle");
         }
     }
+
+    @Reference(
+            name = "claim.config.init.dao",
+            service = ClaimConfigInitDAO.class,
+            cardinality = ReferenceCardinality.OPTIONAL,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unsetClaimConfigInitDAO"
+    )
+    protected void setClaimConfigInitDAO(ClaimConfigInitDAO claimConfigInitDAO) {
+
+        ClaimConfigInitDAO existingDAO = IdentityClaimManagementServiceDataHolder.getInstance().getClaimConfigInitDAO();
+        if (existingDAO != null) {
+            log.warn("Claim config init DAO implementation " + existingDAO.getClass().getName() +
+                    " is registered already and ClaimMetadataStoreFactory is created." +
+                    " So DAO Impl : " + claimConfigInitDAO.getClass().getName() + " will not be registered");
+        } else {
+            IdentityClaimManagementServiceDataHolder.getInstance().setClaimConfigInitDAO(claimConfigInitDAO);
+            if (log.isDebugEnabled()) {
+                log.debug("Claim config init DAO implementation got registered: " +
+                        claimConfigInitDAO.getClass().getName());
+            }
+        }
+    }
+
+    protected void unsetClaimConfigInitDAO(ClaimConfigInitDAO claimConfigInitDAO) {
+
+        IdentityClaimManagementServiceDataHolder.getInstance().setClaimConfigInitDAO(new DefaultClaimConfigInitDAO());
```

Comment:
When unsetting should we set if to default or completely remove it. If we need it shall we improve the debug log to mention the previous set dao was removed and now the default dao is set

Reply:
we remove that DAO implementation and register the default DAO. I have improved the debug logs.